plugin/audit_log/audit_log.c

```diff
@@ -468,8 +488,31 @@ char *audit_log_general_record(char *buf, size_t buflen,
                      "\"%s\",\"%s\",\"%s\",\"%s\",\"%lu\",%d,\"%s\",\"%s\","
                      "\"%s\",\"%s\",\"%s\",\"%s\"\n" };
 
-  query= escape_string(event->general_query, event->general_query_length,
-                       endptr, endbuf - endptr, &endptr, &full_outlen);
+  query_length= system_charset_info->mbmaxlen * event->general_query_length;
+
+  if (query_length < (size_t) (endbuf - endptr))
+  {
+    uint errors;
+    query_length= my_convert(endptr, query_length, system_charset_info,
+                             event->general_query,
+                             event->general_query_length,
+                             event->general_charset, &errors);
```

Comment:
The problem here is that MySQL in reality is using 3-byte subset of UTF-8 internally and `utf8` is the alias of the `utf8mb3`. Here is an example which will give `errors==3`:

```
SET NAMES utf8mb4;
INSERT INTO t VALUES ('𦉘𦟌𦧲');
```

We'll see warning:
```
Warning        1366    Incorrect string value: '\xF0\xA6\x89\x98\xF0\xA6...' for column 'txt' at row 1
```
and ??? in the audit log:

```
"Query","<ID>","<DATETIME>","insert","<CONN_ID>",0,"INSERT INTO t VALUES ('???')","root[root] @ localhost []","localhost","","","ฐานข้อมูล"
"Query","<ID>","<DATETIME>","show_warnings","<CONN_ID>",0,"SHOW WARNINGS","root[root] @ localhost []","localhost","","","ฐานข้อมูล"
```